components/servo_tracing/lib.rs

```diff
@@ -0,0 +1,394 @@
+            Directive::Fields(fields) => tokens.append_all(quote! { #fields }),
+        };
+    }
+}
+
+impl ToTokens for InstrumentConfiguration {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        tokens.append_terminated(&self.0, Punct::new(',', proc_macro2::Spacing::Joint));
+    }
+}
+
+struct InstrumentConfiguration(Vec<Directive>);
+
+impl InstrumentConfiguration {
+    fn inject_servo_profiling(&mut self) -> syn::Result<()> {
+        // establish invariant that self.0 has a Directive::Fields.
+
+        let contains = self.0.iter().any(|a| a.is_fields());
+        if !contains {
+            self.0.push(Directive::Fields(Fields(
+                parse_quote! { fields(servo_profiling = true) },
+            )));
+        }
+        let component = self.0.iter_mut().find(|a| a.is_fields()).unwrap();
+
+        assert!(matches!(component, Directive::Fields(..)));
+        // this will always be Some(fields);
+        let Some(fields) = component.fields_mut() else {
+            unreachable!("We have established that there is always a Directive::fields")
+        };
+
+        fields.inject_servo_profiling()
+    }
+
+    fn inject_level(&mut self) {
+        if self.0.iter().any(|a| a.is_level()) {
+            return;
+        }
+        self.0.push(Directive::Level(parse_quote! { "trace" }));
+    }
+}
+
+impl Parse for InstrumentConfiguration {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let args = Punctuated::<Meta, Token![,]>::parse_terminated(input)?;
+        let mut components = vec![];
+
+        for arg in args {
+            match arg {
+                Meta::List(meta_list) if meta_list.path.is_ident("fields") => {
+                    components.push(Directive::Fields(meta_list.into()));
+                },
+                Meta::NameValue(meta_name_value) if meta_name_value.path.is_ident("level") => {
+                    components.push(Directive::Level(meta_name_value.value));
+                },
+                _ => {
+                    components.push(Directive::Passthrough(arg));
+                },
+            }
+        }
+        Ok(InstrumentConfiguration(components))
+    }
+}
+
+fn instrument_internal(
+    attr: proc_macro2::TokenStream,
+    item: proc_macro2::TokenStream,
+) -> syn::Result<proc_macro2::TokenStream> {
+    // Prepare passthrough arguments for tracing::instrument
+    let mut configuration: InstrumentConfiguration = parse2(attr)?;
+    let input_fn: ItemFn = parse2(item)?;
+
+    configuration.inject_servo_profiling()?;
+    configuration.inject_level();
+
+    let output = quote! {
+        #[cfg_attr(
+            feature = "tracing",
+            tracing::instrument(
+                #configuration
+            )
+        )]
+        #input_fn
+    };
+
+    Ok(output)
+}
+
+#[proc_macro_attribute]
+/// Instruments a function with some sane defaults by automatically:
+///  - setting the attribute behind the "tracing" flag
```

Comment:
Probably wouldn't harm to clarify here, that this implies/requires the using crate has a `tracing` feature flag.